File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/administrativ/X008Mapper.java
package no.nav.melosys.eessi.service.sed.mapper.til_sed.administrativ;

import no.nav.melosys.eessi.controller.dto.SedDataDto;
import no.nav.melosys.eessi.models.SedType;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.nav.Grunn;
import no.nav.melosys.eessi.models.sed.nav.InvalideringSed;
import no.nav.melosys.eessi.models.sed.nav.Sak;
import no.nav.melosys.eessi.models.sed.nav.Ugyldiggjoere;
import no.nav.melosys.eessi.service.sed.mapper.til_sed.SedMapper;

public class X008Mapper implements SedMapper {

    @Override
    public SED mapTilSed(SedDataDto sedData) {
        var sed = SedMapper.super.mapTilSed(sedData);

        sed.getNav().setSak(mapSak(sedData));

        return sed;
    }

    public Sak mapSak(SedDataDto sedData) {
        var sak = new Sak();
        var invalideringSed = new InvalideringSed();

        invalideringSed.setType(sedData.getInvalideringSedDto().getSedTypeSomSkalInvalideres());
        invalideringSed.setUtstedelsesdato(sedData.getInvalideringSedDto().getUtstedelsedato());
        invalideringSed.setGrunn(new Grunn("04", ""));

        sak.setUgyldiggjoere(new Ugyldiggjoere(invalideringSed));

        return sak;
    }

    @Override
    public SedType getSedType() {
        return SedType.X008;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/administrativ/X008Mapper.kt
package no.nav.melosys.eessi.service.sed.mapper.til_sed.administrativ

import no.nav.melosys.eessi.controller.dto.SedDataDto
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.exception.MappingException
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.nav.*
import no.nav.melosys.eessi.service.sed.mapper.til_sed.SedMapper

class X008Mapper : SedMapper {
    override fun getSedType(): SedType = SedType.X008

    override fun mapTilSed(sedData: SedDataDto): SED =
        super.mapTilSed(sedData).also { sed ->
            sed.nav?.let { nav ->
                nav.sak = mapSak(sedData, sed)
            } ?: throw MappingException("nav.sak er påkrevd for X008")
        }

    fun mapSak(sedData: SedDataDto, sed: SED): Sak {
        val invalideringSedDto = sedData.invalideringSedDto
            ?: throw IllegalArgumentException("SedDataDto.invalideringSedDto kan ikke være null")

        return Sak(
            ugyldiggjoere = Ugyldiggjoere(
                sed = InvalideringSed(
                    type = invalideringSedDto.sedTypeSomSkalInvalideres,
                    utstedelsesdato = invalideringSedDto.utstedelsedato,
                    grunn = Grunn("04", "") // 04  = The case was reconsidered and the grounds for the invalidated SED are no longer valid
                )
            ),
            kontekst = mapKontekst(sed)
        )
    }

    private fun mapKontekst(sed: SED) = Kontekst(
        bruker = sed.nav?.bruker ?: throw MappingException("nav.bruker kan ikke være null")
    )
}
